client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -140,7 +140,8 @@ public void readerOffline(String readerId, Position lastPosition) {
         }
 
         return waitForCheckpointComplete(checkpointName, backgroundExecutor)
-                .thenApply(v -> completeCheckpoint(checkpointName));
+                .thenApply(v -> completeCheckpoint(checkpointName))
+                .thenApply(checkpoint -> checkpoint); //Added to prevent users from canceling completeCheckpoint
```

Comment:
Just to confirm that it will still be safe to call `cancel` on the `CompletableFuture` (as the connector does)

Reply:
This change prevents a cancel of the future to skip the `completeCheckpoint`computation, which is the logic that causes the number of outstanding checkpoints to drop. Long story short, yes, this makes it safe to cancel the future and the additional stage will only the `checkpoint` object from being returned. See the test case below for a validation of the expected behavior.

Reply:
May I suggest that a helper method be used to encapsulate this technique, .e.g `Futures.cancellable(...)`.   Would make it easier to identify the locations where cancellation is explicitly supported, and provide overrides in support of taking an action upon cancellation (such as cancelling an upstream future).

Regarding this specific instance, I wonder if the counter should be decremented in exceptional cases too.   Imagine that the caller were to close the `backgroundExecutor`, it would again defeat `completeCheckpoint`.